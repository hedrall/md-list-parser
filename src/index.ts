import MarkdownIt from 'markdown-it';
import * as fs from "fs";
import * as path from "path";

const md = new MarkdownIt();

type ListItem = {
    inline: string,
    body: string[],
    children: ListItem[],
}

type Token = {
    index: number,
    type: string,
    content: string,
    markup: string,
};

type ListNodeTypes = 'paragraph' | 'bullet_list' | 'list_item';
const LIST_NODE_TYPE_COL_NAME_MAP = {
    paragraph: 'paragraphs',
    bullet_list: 'bulletLists',
    list_item: 'listItems',
} as const;

const pickTokensOf = (type: string, tokens: Token[], indexOfOpen: number) => {
    let index = indexOfOpen + 1;
    let count = 1;
    while(count > 0) {
        const token = tokens[index];
        if (token.type ===  `${type}_open`) count++;
        if (token.type ===  `${type}_close`) count--;
        index++;
    }
    const indexOfClose = index - 1;
    return {
        items: tokens.slice(indexOfOpen + 1, indexOfClose),
        indexOfClose,
    };
}

// group by paragraph or bullet_list or list_item
const groupBy = (tokens: Token[]) => {
    const res = {
        paragraphs: [] as Token[][],
        bulletLists: [] as Token[][],
        listItems: [] as Token[][],
    }
    for (let i = 0; i < tokens.length; i++) {
        const token = tokens[i];
        const type = token.type.match(/(paragraph|bullet_list|list_item)_open/)?.[1] as null | ListNodeTypes;

        if (type) {
            const {items, indexOfClose} = pickTokensOf(type, tokens, i);
            res[LIST_NODE_TYPE_COL_NAME_MAP[type]].push(items);
            i = indexOfClose;
        }
    }
    return res;
}

const filterTokensOnlyBulletList = (tokens: Token[]) => {
    const filtered: Token[] = [];

    for (let i = 0; i < tokens.length; i++) {
        const token = tokens[i];

        if (token.type === 'bullet_list_open') {
            const {indexOfClose} = pickTokensOf('bullet_list', tokens, i);
            filtered.push(...tokens.slice(i, indexOfClose + 1));
            i = indexOfClose;
        }
    }

    return filtered;
}

type ParsedList = {
    paragraphs: Token[][],
    bulletLists: ParsedList[],
    listItems: ParsedList[],
}

function getContentOfParagraphs(paragraphs: Token[]) {
    return paragraphs.map(p => p.content).join('\n');
}

type Structured = {
    title: string,
    children?: Structured[] | null,
}
const structure = (parsed: ParsedList): Structured => {
    const type = parsed.bulletLists.length ? 'has-bullet-list' : parsed.listItems.length ? 'has-list-item' : 'has-paragraph';
    if (type === 'has-bullet-list') {
        return {
            title: parsed.paragraphs.map(getContentOfParagraphs).join('\n'),
            children: parsed.bulletLists.flatMap(bl => bl.listItems.map(structure))
        };
    }
    if (type === 'has-list-item') {
        return {
            title: '',
            children: parsed.listItems.map(structure)
        };
    }
    return {
        title: parsed.paragraphs.map(getContentOfParagraphs).join('\n'),
    };
}

const _parse = (tokens: Token[]): ParsedList => {
    const groups = groupBy(tokens);
    return {
        ...groups,
        bulletLists: groups.bulletLists.map(_parse),
        listItems: groups.listItems.map(_parse),
    }
}

export type Parsed = Structured;
export const parse = (input: string) => {
    // tokenize
    const tokens = md.parse(input, {}).map((token, index) => {
        return {
            index,
            type: token.type,
            content: token.content,
            markup: token.markup,
        };
    });

    const filteredTokens = filterTokensOnlyBulletList(tokens);
    // parse to ast like
    const parsed = _parse(filteredTokens);

    // modified to structured
    return structure(parsed).children || undefined;
};
