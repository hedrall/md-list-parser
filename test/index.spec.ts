import { parse } from '../src/index';

test('README case', () => {
    const input = `
- a
  - b
  - d
    - e
- f
`;
    expect(parse(input)).toMatchSnapshot();
});

test('main case', () => {
    const input = `
- (4) a
    - a-body
- (4) b
    - b-body
    - (2) b1
        - (1) b1-1
            - (b1-1-body)
        - (1) b1-2
            - b1-2-body
    - (2) b2
        - b2-body
    - b-body-2
`;
    expect(parse(input)).toMatchSnapshot();
});

test('mixed with other markdown', () => {
    const input = `
    
inline 1

# heading1
## heading2

\`\`\`
    code block
\`\`\`


- (4) a
    - a-body
- (4) b
    - b-body
    - (2) b1
        - (1) b1-1
            - (b1-1-body)
        - (1) b1-2
            - b1-2-body
    - (2) b2
        - b2-body
    - b-body-2
    
    
### heading3

inline 2
`;
    expect(parse(input)).toMatchSnapshot();
});

test('two lists', () => {
    const input = `
# one

- 1
    - 2
    - 3

# two

- a
- b
    - c 
`;
    expect(parse(input)).toMatchSnapshot();
});

test('+ list', () => {
    const input = `
+ 1
    + 2
    + 3
`;
    expect(parse(input)).toMatchSnapshot();
});

test('num list', () => {
    const input = `
1. a
1. b
    1. c
`;
    expect(parse(input)).toBe(undefined);
});
