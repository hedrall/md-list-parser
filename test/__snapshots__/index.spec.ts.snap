// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`+ list 1`] = `
[
  {
    "children": [
      {
        "title": "2",
      },
      {
        "title": "3",
      },
    ],
    "title": "1",
  },
]
`;

exports[`README case 1`] = `
[
  {
    "children": [
      {
        "title": "b",
      },
      {
        "children": [
          {
            "title": "e",
          },
        ],
        "title": "d",
      },
    ],
    "title": "a",
  },
  {
    "title": "f",
  },
]
`;

exports[`main case 1`] = `
[
  {
    "children": [
      {
        "title": "a-body",
      },
    ],
    "title": "(4) a",
  },
  {
    "children": [
      {
        "title": "b-body",
      },
      {
        "children": [
          {
            "children": [
              {
                "title": "(b1-1-body)",
              },
            ],
            "title": "(1) b1-1",
          },
          {
            "children": [
              {
                "title": "b1-2-body",
              },
            ],
            "title": "(1) b1-2",
          },
        ],
        "title": "(2) b1",
      },
      {
        "children": [
          {
            "title": "b2-body",
          },
        ],
        "title": "(2) b2",
      },
      {
        "title": "b-body-2",
      },
    ],
    "title": "(4) b",
  },
]
`;

exports[`mixed with other markdown 1`] = `
[
  {
    "children": [
      {
        "title": "a-body",
      },
    ],
    "title": "(4) a",
  },
  {
    "children": [
      {
        "title": "b-body",
      },
      {
        "children": [
          {
            "children": [
              {
                "title": "(b1-1-body)",
              },
            ],
            "title": "(1) b1-1",
          },
          {
            "children": [
              {
                "title": "b1-2-body",
              },
            ],
            "title": "(1) b1-2",
          },
        ],
        "title": "(2) b1",
      },
      {
        "children": [
          {
            "title": "b2-body",
          },
        ],
        "title": "(2) b2",
      },
      {
        "title": "b-body-2",
      },
    ],
    "title": "(4) b",
  },
]
`;

exports[`two lists 1`] = `
[
  {
    "children": [
      {
        "title": "2",
      },
      {
        "title": "3",
      },
    ],
    "title": "1",
  },
  {
    "title": "a",
  },
  {
    "children": [
      {
        "title": "c",
      },
    ],
    "title": "b",
  },
]
`;
